// @guildedjs/api/dist/generated/router/services/AnnouncementCommentsService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnnouncementCommentsService = void 0;
class AnnouncementCommentsService {
    constructor(httpRequest) {
        this.httpRequest = httpRequest;
    }
    /**
     * Create a comment on an announcement
     * @returns any Success
     * @throws ApiError
     */
    announcementCommentCreate({ channelId, announcementId, requestBody, }) {
        return this.httpRequest.request({
            method: "POST",
            url: "/channels/{channelId}/announcements/{announcementId}/comments",
            path: {
                channelId: channelId,
                announcementId: announcementId,
            },
            body: requestBody,
            mediaType: "application/json",
        });
    }
    /**
     * Get an announcement's comments
     * @returns any Success
     * @throws ApiError
     */
    announcementCommentReadMany({ channelId, announcementId }) {
        return this.httpRequest.request({
            method: "GET",
            url: "/channels/{channelId}/announcements/{announcementId}/comments",
            path: {
                channelId: channelId,
                announcementId: announcementId,
            },
        });
    }
    /**
     * Get a comment on the announcement
     * @returns any Success
     * @throws ApiError
     */
    announcementCommentRead({ channelId, announcementId, announcementCommentId }) {
        return this.httpRequest.request({
            method: "GET",
            url: "/channels/{channelId}/announcements/{announcementId}/comments/{announcementCommentId}",
            path: {
                channelId: channelId,
                announcementId: announcementId,
                announcementCommentId: announcementCommentId,
            },
        });
    }
    /**
     * Update an announcement comment
     * @returns any Success
     * @throws ApiError
     */
    announcementCommentUpdate({ channelId, announcementId, announcementCommentId, requestBody, }) {
        return this.httpRequest.request({
            method: "PATCH",
            url: "/channels/{channelId}/announcements/{announcementId}/comments/{announcementCommentId}",
            path: {
                channelId: channelId,
                announcementId: announcementId,
                announcementCommentId: announcementCommentId,
            },
            body: requestBody,
            mediaType: "application/json",
        });
    }
    /**
     * Delete an announcement comment
     * @returns void
     * @throws ApiError
     */
    announcementCommentDelete({ channelId, announcementId, announcementCommentId }) {
        return this.httpRequest.request({
            method: "DELETE",
            url: "/channels/{channelId}/announcements/{announcementId}/comments/{announcementCommentId}",
            path: {
                channelId: channelId,
                announcementId: announcementId,
                announcementCommentId: announcementCommentId,
            },
        });
    }
}
exports.AnnouncementCommentsService = AnnouncementCommentsService;
