// @guildedjs/api/dist/generated/router/services/CalendarEventSeriesService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CalendarEventSeriesService = void 0;
class CalendarEventSeriesService {
    constructor(httpRequest) {
        this.httpRequest = httpRequest;
    }
    /**
     * Update a calendar event series
     * @returns void
     * @throws ApiError
     */
    calendarEventSeriesUpdate({ channelId, calendarEventSeriesId, requestBody, }) {
        return this.httpRequest.request({
            method: "PATCH",
            url: "/channels/{channelId}/event_series/{calendarEventSeriesId}",
            path: {
                channelId: channelId,
                calendarEventSeriesId: calendarEventSeriesId,
            },
            body: requestBody,
            mediaType: "application/json",
        });
    }
    /**
     * Delete a calendar event series
     * @returns void
     * @throws ApiError
     */
    calendarEventSeriesDelete({ channelId, calendarEventSeriesId, requestBody, }) {
        return this.httpRequest.request({
            method: "DELETE",
            url: "/channels/{channelId}/event_series/{calendarEventSeriesId}",
            path: {
                channelId: channelId,
                calendarEventSeriesId: calendarEventSeriesId,
            },
            body: requestBody,
            mediaType: "application/json",
        });
    }
}
exports.CalendarEventSeriesService = CalendarEventSeriesService;
