// @guildedjs/api/dist/generated/router/services/CalendarEventCommentsService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CalendarEventCommentsService = void 0;
class CalendarEventCommentsService {
    constructor(httpRequest) {
        this.httpRequest = httpRequest;
    }
    /**
     * Create a comment on an event
     * @returns any Success
     * @throws ApiError
     */
    calendarEventCommentCreate({ channelId, calendarEventId, requestBody, }) {
        return this.httpRequest.request({
            method: "POST",
            url: "/channels/{channelId}/events/{calendarEventId}/comments",
            path: {
                channelId: channelId,
                calendarEventId: calendarEventId,
            },
            body: requestBody,
            mediaType: "application/json",
        });
    }
    /**
     * Get a calendar event's comments
     * @returns any Success
     * @throws ApiError
     */
    calendarEventCommentReadMany({ channelId, calendarEventId }) {
        return this.httpRequest.request({
            method: "GET",
            url: "/channels/{channelId}/events/{calendarEventId}/comments",
            path: {
                channelId: channelId,
                calendarEventId: calendarEventId,
            },
        });
    }
    /**
     * Get a comment on the calendar event
     * @returns any Success
     * @throws ApiError
     */
    calendarEventCommentRead({ channelId, calendarEventId, calendarEventCommentId }) {
        return this.httpRequest.request({
            method: "GET",
            url: "/channels/{channelId}/events/{calendarEventId}/comments/{calendarEventCommentId}",
            path: {
                channelId: channelId,
                calendarEventId: calendarEventId,
                calendarEventCommentId: calendarEventCommentId,
            },
        });
    }
    /**
     * Update a calendar event comment
     * @returns any Success
     * @throws ApiError
     */
    calendarEventCommentUpdate({ channelId, calendarEventId, calendarEventCommentId, requestBody, }) {
        return this.httpRequest.request({
            method: "PATCH",
            url: "/channels/{channelId}/events/{calendarEventId}/comments/{calendarEventCommentId}",
            path: {
                channelId: channelId,
                calendarEventId: calendarEventId,
                calendarEventCommentId: calendarEventCommentId,
            },
            body: requestBody,
            mediaType: "application/json",
        });
    }
    /**
     * Delete a calendar event comment
     * @returns void
     * @throws ApiError
     */
    calendarEventCommentDelete({ channelId, calendarEventId, calendarEventCommentId }) {
        return this.httpRequest.request({
            method: "DELETE",
            url: "/channels/{channelId}/events/{calendarEventId}/comments/{calendarEventCommentId}",
            path: {
                channelId: channelId,
                calendarEventId: calendarEventId,
                calendarEventCommentId: calendarEventCommentId,
            },
        });
    }
}
exports.CalendarEventCommentsService = CalendarEventCommentsService;
