// @guildedjs/api/dist/generated/router/services/CategoriesService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CategoriesService = void 0;
class CategoriesService {
    constructor(httpRequest) {
        this.httpRequest = httpRequest;
    }
    /**
     * Create a category
     * @returns any Success
     * @throws ApiError
     */
    categoryCreate({ serverId, requestBody, }) {
        return this.httpRequest.request({
            method: "POST",
            url: "/servers/{serverId}/categories",
            path: {
                serverId: serverId,
            },
            body: requestBody,
            mediaType: "application/json",
        });
    }
    /**
     * Read a category
     * @returns any Success
     * @throws ApiError
     */
    categoryRead({ serverId, categoryId }) {
        return this.httpRequest.request({
            method: "GET",
            url: "/servers/{serverId}/categories/{categoryId}",
            path: {
                serverId: serverId,
                categoryId: categoryId,
            },
        });
    }
    /**
     * Update a category
     * @returns any Success
     * @throws ApiError
     */
    categoryUpdate({ serverId, categoryId, requestBody, }) {
        return this.httpRequest.request({
            method: "PATCH",
            url: "/servers/{serverId}/categories/{categoryId}",
            path: {
                serverId: serverId,
                categoryId: categoryId,
            },
            body: requestBody,
            mediaType: "application/json",
        });
    }
    /**
     * Delete a category
     * @returns any Success
     * @throws ApiError
     */
    categoryDelete({ serverId, categoryId }) {
        return this.httpRequest.request({
            method: "DELETE",
            url: "/servers/{serverId}/categories/{categoryId}",
            path: {
                serverId: serverId,
                categoryId: categoryId,
            },
        });
    }
}
exports.CategoriesService = CategoriesService;
