// @guildedjs/api/dist/generated/router/services/ForumCommentsService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForumCommentsService = void 0;
class ForumCommentsService {
    constructor(httpRequest) {
        this.httpRequest = httpRequest;
    }
    /**
     * Create a forum topic comment
     * @returns any Success
     * @throws ApiError
     */
    forumTopicCommentCreate({ channelId, forumTopicId, requestBody, }) {
        return this.httpRequest.request({
            method: "POST",
            url: "/channels/{channelId}/topics/{forumTopicId}/comments",
            path: {
                channelId: channelId,
                forumTopicId: forumTopicId,
            },
            body: requestBody,
            mediaType: "application/json",
        });
    }
    /**
     * Get a forum topic's comments
     * @returns any Success
     * @throws ApiError
     */
    forumTopicCommentReadMany({ channelId, forumTopicId }) {
        return this.httpRequest.request({
            method: "GET",
            url: "/channels/{channelId}/topics/{forumTopicId}/comments",
            path: {
                channelId: channelId,
                forumTopicId: forumTopicId,
            },
        });
    }
    /**
     * Get a comment on a forum topic
     * @returns any Success
     * @throws ApiError
     */
    forumTopicCommentRead({ channelId, forumTopicId, forumTopicCommentId }) {
        return this.httpRequest.request({
            method: "GET",
            url: "/channels/{channelId}/topics/{forumTopicId}/comments/{forumTopicCommentId}",
            path: {
                channelId: channelId,
                forumTopicId: forumTopicId,
                forumTopicCommentId: forumTopicCommentId,
            },
        });
    }
    /**
     * Update a forum topic comment
     * @returns any Success
     * @throws ApiError
     */
    forumTopicCommentUpdate({ channelId, forumTopicId, forumTopicCommentId, requestBody, }) {
        return this.httpRequest.request({
            method: "PATCH",
            url: "/channels/{channelId}/topics/{forumTopicId}/comments/{forumTopicCommentId}",
            path: {
                channelId: channelId,
                forumTopicId: forumTopicId,
                forumTopicCommentId: forumTopicCommentId,
            },
            body: requestBody,
            mediaType: "application/json",
        });
    }
    /**
     * Delete a forum topic comment
     * @returns void
     * @throws ApiError
     */
    forumTopicCommentDelete({ channelId, forumTopicId, forumTopicCommentId }) {
        return this.httpRequest.request({
            method: "DELETE",
            url: "/channels/{channelId}/topics/{forumTopicId}/comments/{forumTopicCommentId}",
            path: {
                channelId: channelId,
                forumTopicId: forumTopicId,
                forumTopicCommentId: forumTopicCommentId,
            },
        });
    }
}
exports.ForumCommentsService = ForumCommentsService;
