// @guildedjs/api/dist/generated/router/services/GroupMembershipService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GroupMembershipService = void 0;
class GroupMembershipService {
    constructor(httpRequest) {
        this.httpRequest = httpRequest;
    }
    /**
     * Add member to group
     * @returns void
     * @throws ApiError
     */
    groupMembershipCreate({ groupId, userId, }) {
        return this.httpRequest.request({
            method: "PUT",
            url: "/groups/{groupId}/members/{userId}",
            path: {
                groupId: groupId,
                userId: userId,
            },
        });
    }
    /**
     * Remove member from group
     * @returns void
     * @throws ApiError
     */
    groupMembershipDelete({ groupId, userId, }) {
        return this.httpRequest.request({
            method: "DELETE",
            url: "/groups/{groupId}/members/{userId}",
            path: {
                groupId: groupId,
                userId: userId,
            },
        });
    }
}
exports.GroupMembershipService = GroupMembershipService;
