// @guildedjs/api/dist/generated/router/services/ListItemsService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListItemsService = void 0;
class ListItemsService {
    constructor(httpRequest) {
        this.httpRequest = httpRequest;
    }
    /**
     * Create a list item
     * @returns any Success
     * @throws ApiError
     */
    listItemCreate({ channelId, requestBody, }) {
        return this.httpRequest.request({
            method: "POST",
            url: "/channels/{channelId}/items",
            path: {
                channelId: channelId,
            },
            body: requestBody,
            mediaType: "application/json",
        });
    }
    /**
     * Get list items within a channel
     * @returns any Success
     * @throws ApiError
     */
    listItemReadMany({ channelId }) {
        return this.httpRequest.request({
            method: "GET",
            url: "/channels/{channelId}/items",
            path: {
                channelId: channelId,
            },
        });
    }
    /**
     * Get a list item
     * @returns any Success
     * @throws ApiError
     */
    listItemRead({ channelId, listItemId }) {
        return this.httpRequest.request({
            method: "GET",
            url: "/channels/{channelId}/items/{listItemId}",
            path: {
                channelId: channelId,
                listItemId: listItemId,
            },
        });
    }
    /**
     * Update a list item
     * @returns any Success
     * @throws ApiError
     */
    listItemUpdate({ channelId, listItemId, requestBody, }) {
        return this.httpRequest.request({
            method: "PATCH",
            url: "/channels/{channelId}/items/{listItemId}",
            path: {
                channelId: channelId,
                listItemId: listItemId,
            },
            body: requestBody,
            mediaType: "application/json",
        });
    }
    /**
     * Delete a list item
     * @returns void
     * @throws ApiError
     */
    listItemDelete({ channelId, listItemId }) {
        return this.httpRequest.request({
            method: "DELETE",
            url: "/channels/{channelId}/items/{listItemId}",
            path: {
                channelId: channelId,
                listItemId: listItemId,
            },
        });
    }
    /**
     * Complete a list item
     * @returns void
     * @throws ApiError
     */
    listItemCompleteCreate({ channelId, listItemId }) {
        return this.httpRequest.request({
            method: "POST",
            url: "/channels/{channelId}/items/{listItemId}/complete",
            path: {
                channelId: channelId,
                listItemId: listItemId,
            },
        });
    }
    /**
     * Uncomplete a list item
     * @returns void
     * @throws ApiError
     */
    listItemCompleteDelete({ channelId, listItemId }) {
        return this.httpRequest.request({
            method: "DELETE",
            url: "/channels/{channelId}/items/{listItemId}/complete",
            path: {
                channelId: channelId,
                listItemId: listItemId,
            },
        });
    }
}
exports.ListItemsService = ListItemsService;
