// @guildedjs/api/dist/generated/router/services/MemberBansService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MemberBansService = void 0;
class MemberBansService {
    constructor(httpRequest) {
        this.httpRequest = httpRequest;
    }
    /**
     * Create a server ban
     * Also known as banning a server member
     * @returns any Success
     * @throws ApiError
     */
    serverMemberBanCreate({ serverId, userId, requestBody, }) {
        return this.httpRequest.request({
            method: "POST",
            url: "/servers/{serverId}/bans/{userId}",
            path: {
                serverId: serverId,
                userId: userId,
            },
            body: requestBody,
            mediaType: "application/json",
        });
    }
    /**
     * Get a server ban
     * @returns any Success
     * @throws ApiError
     */
    serverMemberBanRead({ serverId, userId, }) {
        return this.httpRequest.request({
            method: "GET",
            url: "/servers/{serverId}/bans/{userId}",
            path: {
                serverId: serverId,
                userId: userId,
            },
        });
    }
    /**
     * Delete a server ban
     * Also known as unbanning a server member
     * @returns void
     * @throws ApiError
     */
    serverMemberBanDelete({ serverId, userId, }) {
        return this.httpRequest.request({
            method: "DELETE",
            url: "/servers/{serverId}/bans/{userId}",
            path: {
                serverId: serverId,
                userId: userId,
            },
        });
    }
    /**
     * Get server bans
     * @returns any Success
     * @throws ApiError
     */
    serverMemberBanReadMany({ serverId }) {
        return this.httpRequest.request({
            method: "GET",
            url: "/servers/{serverId}/bans",
            path: {
                serverId: serverId,
            },
        });
    }
}
exports.MemberBansService = MemberBansService;
