// @guildedjs/api/dist/generated/router/services/RolesService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RolesService = void 0;
class RolesService {
    constructor(httpRequest) {
        this.httpRequest = httpRequest;
    }
    /**
     * Get a role
     * @returns any Success
     * @throws ApiError
     */
    roleRead({ serverId, roleId }) {
        return this.httpRequest.request({
            method: "GET",
            url: "/servers/{serverId}/roles/{roleId}",
            path: {
                serverId: serverId,
                roleId: roleId,
            },
        });
    }
    /**
     * Delete a role
     * @returns any Success
     * @throws ApiError
     */
    roleDelete({ serverId, roleId }) {
        return this.httpRequest.request({
            method: "DELETE",
            url: "/servers/{serverId}/roles/{roleId}",
            path: {
                serverId: serverId,
                roleId: roleId,
            },
        });
    }
    /**
     * Update a role
     * @returns any Success
     * @throws ApiError
     */
    roleUpdate({ serverId, roleId, requestBody, }) {
        return this.httpRequest.request({
            method: "PATCH",
            url: "/servers/{serverId}/roles/{roleId}",
            path: {
                serverId: serverId,
                roleId: roleId,
            },
            body: requestBody,
            mediaType: "application/json",
        });
    }
    /**
     * Get roles
     * @returns any Success
     * @throws ApiError
     */
    roleReadMany({ serverId }) {
        return this.httpRequest.request({
            method: "GET",
            url: "/servers/{serverId}/roles",
            path: {
                serverId: serverId,
            },
        });
    }
    /**
     * Create a role
     * @returns any Success
     * @throws ApiError
     */
    roleCreate({ serverId, requestBody, }) {
        return this.httpRequest.request({
            method: "POST",
            url: "/servers/{serverId}/roles",
            path: {
                serverId: serverId,
            },
            body: requestBody,
            mediaType: "application/json",
        });
    }
    /**
     * Update a role permission
     * @returns any Success
     * @throws ApiError
     */
    rolePermissionUpdate({ serverId, roleId, requestBody, }) {
        return this.httpRequest.request({
            method: "PATCH",
            url: "/servers/{serverId}/roles/{roleId}/permissions",
            path: {
                serverId: serverId,
                roleId: roleId,
            },
            body: requestBody,
            mediaType: "application/json",
        });
    }
    /**
     * Read a server member's permissions
     * If the user has *all* of the permissions passed, the HTTP status code will be 200; if user is missing one or more permissions passed, the HTTP status code will be [418](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/418) with a `meta` property containing `missingPermissions` of the permissions the user does not have (this is a 418 to differentiate from a 403 if the _bot_ did not have permissions to perform the request, not the _user_). Tip: Use the `HEAD` HTTP method for this route if you only care about if the user has permissions and not the response body containing what's missing.
     * @returns any Success
     * @throws ApiError
     */
    serverMemberPermissionsRead({ serverId, userId, ids, }) {
        return this.httpRequest.request({
            method: "GET",
            url: "/servers/{serverId}/members/{userId}/permissions",
            path: {
                serverId: serverId,
                userId: userId,
            },
            query: {
                ids: ids,
            },
        });
    }
}
exports.RolesService = RolesService;
