// @guildedjs/api/dist/generated/router/services/RoleMembershipService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RoleMembershipService = void 0;
class RoleMembershipService {
    constructor(httpRequest) {
        this.httpRequest = httpRequest;
    }
    /**
     * Assign role to member
     * @returns void
     * @throws ApiError
     */
    roleMembershipCreate({ serverId, userId, roleId, }) {
        return this.httpRequest.request({
            method: "PUT",
            url: "/servers/{serverId}/members/{userId}/roles/{roleId}",
            path: {
                serverId: serverId,
                userId: userId,
                roleId: roleId,
            },
        });
    }
    /**
     * Remove role from member
     * @returns void
     * @throws ApiError
     */
    roleMembershipDelete({ serverId, userId, roleId, }) {
        return this.httpRequest.request({
            method: "DELETE",
            url: "/servers/{serverId}/members/{userId}/roles/{roleId}",
            path: {
                serverId: serverId,
                userId: userId,
                roleId: roleId,
            },
        });
    }
    /**
     * Get member roles
     * Get a list of the roles assigned to a member
     * @returns any Success
     * @throws ApiError
     */
    roleMembershipReadMany({ serverId, userId, }) {
        return this.httpRequest.request({
            method: "GET",
            url: "/servers/{serverId}/members/{userId}/roles",
            path: {
                serverId: serverId,
                userId: userId,
            },
        });
    }
}
exports.RoleMembershipService = RoleMembershipService;
