// @guildedjs/api/dist/generated/router/services/ServerSubscriptionService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerSubscriptionService = void 0;
class ServerSubscriptionService {
    constructor(httpRequest) {
        this.httpRequest = httpRequest;
    }
    /**
     * Get a server subscription tier
     * @returns any Success
     * @throws ApiError
     */
    serverSubscriptionTierRead({ serverId, serverSubscriptionTierType }) {
        return this.httpRequest.request({
            method: "GET",
            url: "/servers/{serverId}/subscriptions/tiers/{serverSubscriptionTierType}",
            path: {
                serverId: serverId,
                serverSubscriptionTierType: serverSubscriptionTierType,
            },
        });
    }
    /**
     * Get server subscription tiers
     * @returns any Success
     * @throws ApiError
     */
    serverSubscriptionTierReadMany({ serverId }) {
        return this.httpRequest.request({
            method: "GET",
            url: "/servers/{serverId}/subscriptions/tiers",
            path: {
                serverId: serverId,
            },
        });
    }
}
exports.ServerSubscriptionService = ServerSubscriptionService;
