// @guildedjs/api/dist/generated/router/services/UserStatusService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserStatusService = void 0;
class UserStatusService {
    constructor(httpRequest) {
        this.httpRequest = httpRequest;
    }
    /**
     * Update your status
     * This value can be a max of 24 hours into the futre. If you are wanting a longer expiration, set this value to null so that it never expires. **Note** - at this time, you can only set a status on your own user
     * @returns void
     * @throws ApiError
     */
    userStatusCreate({ userId, requestBody, }) {
        return this.httpRequest.request({
            method: "PUT",
            url: "/users/{userId}/status",
            path: {
                userId: userId,
            },
            body: requestBody,
            mediaType: "application/json",
        });
    }
    /**
     * Delete your status
     * **Note** - at this time, you can only set a status on your own user
     * @returns void
     * @throws ApiError
     */
    userStatusDelete({ userId }) {
        return this.httpRequest.request({
            method: "DELETE",
            url: "/users/{userId}/status",
            path: {
                userId: userId,
            },
        });
    }
}
exports.UserStatusService = UserStatusService;
