// @guildedjs/api/dist/generated/router/services/UsersService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UsersService = void 0;
class UsersService {
    constructor(httpRequest) {
        this.httpRequest = httpRequest;
    }
    /**
     * Get a user
     * **Note** - at this time, you can only retrieve your own user
     * @returns any Success
     * @throws ApiError
     */
    userRead({ userId }) {
        return this.httpRequest.request({
            method: "GET",
            url: "/users/{userId}",
            path: {
                userId: userId,
            },
        });
    }
    /**
     * Get a users servers
     * **Note** - at this time, you can only retrieve your own servers
     * @returns any Success
     * @throws ApiError
     */
    userServerReadMany({ userId }) {
        return this.httpRequest.request({
            method: "GET",
            url: "/users/{userId}/servers",
            path: {
                userId: userId,
            },
        });
    }
}
exports.UsersService = UsersService;
