// @guildedjs/api/dist/rest/util/routes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DOMAINS = exports.ROUTES = void 0;
exports.ROUTES = {
    // Channel Endpoints
    channels: () => `/channels`,
    channel: (channelId) => `/channels/${channelId}`,
    // Channel Endpoints
    channelMessages: (channelId) => `/channels/${channelId}/messages`,
    channelMessage: (channelId, messageId) => `/channels/${channelId}/messages/${messageId}`,
    // CalendarEvent Endpoints
    calendarEvents: (channelId) => `/channels/${channelId}/events`,
    calendarEvent: (channelId, calendarEventId) => `/channels/${channelId}/events/${calendarEventId}`,
    // CalendarRsvp Endpoints
    calendarEventRsvps: (channelId, calendarEventId) => `/channels/${channelId}/events/${calendarEventId}/rsvps`,
    calendarEventRsvp: (channelId, calendarEventId, userId) => `/channels/${channelId}/events/${calendarEventId}/rsvps/${userId}`,
    calendarEventRsvpsMany: (channelId, calendarEventId) => `/channels/${channelId}/events/${calendarEventId}/rsvps`,
    // Server Endpoints
    server: (serverId) => `/servers/${serverId}`,
    // Member Endpoints
    memberNickname: (serverId, userId) => `/servers/${serverId}/members/${userId}/nickname`,
    memberRoles: (serverId, userId) => `/servers/${serverId}/members/${userId}/roles`,
    member: (serverId, userId) => `/servers/${serverId}/members/${userId}`,
    members: (serverId) => `/servers/${serverId}/members`,
    memberBan: (serverId, userId) => `/servers/${serverId}/bans/${userId}`,
    memberBans: (serverId) => `/servers/${serverId}/bans`,
    // User Endpoints
    user: (userId) => `/users/${userId}`,
    me: () => `/users/@me`,
    // Forum Endpoints
    forumTopics: (channelId) => `/channels/${channelId}/topics`,
    forumTopic: (channelId, forumTopicId) => `/channels/${channelId}/topics/${forumTopicId}`,
    forumTopicPin: (channelId, forumTopicId) => `/channels/${channelId}/topics/${forumTopicId}/pin`,
    forumTopicLock: (channelId, forumTopicId) => `/channels/${channelId}/topics/${forumTopicId}/lock`,
    // List Endpoints
    listItems: (channelId) => `/channels/${channelId}/items`,
    listItem: (channelId, itemId) => `/channels/${channelId}/items/${itemId}`,
    listItemComplete: (channelId, itemId) => `/channels/${channelId}/items/${itemId}/complete`,
    // Docs Endpoints
    channelDocs: (channelId) => `/channels/${channelId}/docs`,
    channelDoc: (channelId, docId) => `/channels/${channelId}/docs/${docId}`,
    // Reactions Endpoints
    channelReaction: (channelId, contentId, emoteId) => `/channels/${channelId}/content/${contentId}/emotes/${emoteId}`,
    // Server XP Endpoints
    memberXP: (serverId, userId) => `/servers/${serverId}/members/${userId}/xp`,
    roleXP: (serverId, roleId) => `/servers/${serverId}/roles/${roleId}/xp`,
    // Social Links Endpoints
    memberSocialLinks: (serverId, userId, type) => `/servers/${serverId}/members/${userId}/social-links/${type}`,
    // Group Memberships Endpoints
    groupMember: (groupId, userId) => `/groups/${groupId}/members/${userId}`,
    // Role Memberships Endpoints
    memberRole: (serverId, memberId, roleId) => `/servers/${serverId}/members/${memberId}/roles/${roleId}`,
    // Webhook Endpoints
    serverWebhooks: (serverId) => `/servers/${serverId}/webhooks`,
    serverWebhook: (serverId, webhookId) => `/servers/${serverId}/webhooks/${webhookId}`,
};
exports.DOMAINS = {
    AWS_CDN: "https://s3-us-west-2.amazonaws.com/www.guilded.gg/",
    BASE_DOMAIN: "www.guilded.gg",
    API_DOMAIN: "api.guilded.gg",
    CDN: "img.guildedcdn.com",
    IMAGE_CDN_DOMAIN: "img.guildedcdn.com",
    MEDIA_DOMAIN: "media.guilded.gg",
};
