// @guildedjs/api/dist/generated/api-typings/v1/events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebSocketEvents = exports.WSOpCodes = void 0;
var WSOpCodes;
(function (WSOpCodes) {
    WSOpCodes[WSOpCodes["SUCCESS"] = 0] = "SUCCESS";
    WSOpCodes[WSOpCodes["WELCOME"] = 1] = "WELCOME";
    WSOpCodes[WSOpCodes["RESUME"] = 2] = "RESUME";
    WSOpCodes[WSOpCodes["ERROR"] = 8] = "ERROR";
    WSOpCodes[WSOpCodes["PING"] = 9] = "PING";
    WSOpCodes[WSOpCodes["PONG"] = 10] = "PONG";
})(WSOpCodes = exports.WSOpCodes || (exports.WSOpCodes = {}));
exports.WebSocketEvents = {
    CalendarEventCreated: "CalendarEventCreated",
    CalendarEventUpdated: "CalendarEventUpdated",
    CalendarEventDeleted: "CalendarEventDeleted",
    CalendarEventRsvpUpdated: "CalendarEventRsvpUpdated",
    CalendarEventRsvpManyUpdated: "CalendarEventRsvpManyUpdated",
    CalendarEventRsvpDeleted: "CalendarEventRsvpDeleted",
    ChatMessageCreated: "ChatMessageCreated",
    ChatMessageUpdated: "ChatMessageUpdated",
    ChatMessageDeleted: "ChatMessageDeleted",
    ServerMemberJoined: "ServerMemberJoined",
    ServerMemberRemoved: "ServerMemberRemoved",
    ServerMemberUpdated: "ServerMemberUpdated",
    ServerMemberBanned: "ServerMemberBanned",
    ServerMemberUnbanned: "ServerMemberUnbanned",
    ServerMemberSocialLinkCreated: "ServerMemberSocialLinkCreated",
    ServerMemberSocialLinkUpdated: "ServerMemberSocialLinkUpdated",
    ServerMemberSocialLinkDeleted: "ServerMemberSocialLinkDeleted",
    BotServerMembershipCreated: "BotServerMembershipCreated",
    BotServerMembershipDeleted: "BotServerMembershipDeleted",
    ServerRolesUpdated: "ServerRolesUpdated",
    ServerWebhookCreated: "ServerWebhookCreated",
    ServerWebhookUpdated: "ServerWebhookUpdated",
    ListItemCompleted: "ListItemCompleted",
    ListItemUncompleted: "ListItemUncompleted",
    ListItemCreated: "ListItemCreated",
    ListItemUpdated: "ListItemUpdated",
    ListItemDeleted: "ListItemDeleted",
    ServerChannelCreated: "ServerChannelCreated",
    ServerChannelUpdated: "ServerChannelUpdated",
    ServerChannelDeleted: "ServerChannelDeleted",
    DocCreated: "DocCreated",
    DocUpdated: "DocUpdated",
    DocDeleted: "DocDeleted",
    ChannelMessageReactionCreated: "ChannelMessageReactionCreated",
    ChannelMessageReactionDeleted: "ChannelMessageReactionDeleted",
    ForumTopicCreated: "ForumTopicCreated",
    ForumTopicUpdated: "ForumTopicUpdated",
    ForumTopicDeleted: "ForumTopicDeleted",
    ForumTopicPinned: "ForumTopicPinned",
    ForumTopicUnpinned: "ForumTopicUnpinned",
    ForumTopicLocked: "ForumTopicLocked",
    ForumTopicUnlocked: "ForumTopicUnlocked",
};
