// guilded.js/dist/gateway/ClientGatewayHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientGatewayHandler = void 0;
const api_1 = require("@guildedjs/api");
const BotEventHandler_1 = require("./handler/BotEventHandler");
const CalendarEventHandler_1 = require("./handler/CalendarEventHandler");
const DocEventHandler_1 = require("./handler/DocEventHandler");
const ForumEventHandler_1 = require("./handler/ForumEventHandler");
const ListEventHandler_1 = require("./handler/ListEventHandler");
const MessageEventHandler_1 = require("./handler/MessageEventHandler");
const ReactionEventHandler_1 = require("./handler/ReactionEventHandler");
const ServerChannelEventHandler_1 = require("./handler/ServerChannelEventHandler");
const ServerEventHandler_1 = require("./handler/ServerEventHandler");
const ServerMemberEventHandler_1 = require("./handler/ServerMemberEventHandler");
const ServerWebhookEventHandler_1 = require("./handler/ServerWebhookEventHandler");
class ClientGatewayHandler {
    constructor(client) {
        this.client = client;
        this.calendarEventHandler = new CalendarEventHandler_1.CalendarEventHandler(this.client);
        this.calendarEventRsvpHandler = new CalendarEventHandler_1.CalendarEventRsvpHandler(this.client);
        this.messageHandler = new MessageEventHandler_1.MessageEventHandler(this.client);
        this.ServerHandler = new ServerEventHandler_1.ServerEventHandler(this.client);
        this.ServerMemberHandler = new ServerMemberEventHandler_1.ServerMemberEventHandler(this.client);
        this.ServerWebhookHandler = new ServerWebhookEventHandler_1.ServerWebhookEventHandler(this.client);
        this.listHandler = new ListEventHandler_1.ListEventHandler(this.client);
        this.ServerChannelHandler = new ServerChannelEventHandler_1.ServerChannelEventHandler(this.client);
        this.docHandler = new DocEventHandler_1.DocEventHandler(this.client);
        this.reactionHandler = new ReactionEventHandler_1.ReactionEventHandler(this.client);
        this.forumHandler = new ForumEventHandler_1.ForumEventHandler(this.client);
        this.botHandler = new BotEventHandler_1.BotEventHandler(this.client);
        this.eventToHandlerMap = {
            [api_1.WebSocketEvents.CalendarEventCreated]: (data) => this.calendarEventHandler.calendarEventCreated(data),
            [api_1.WebSocketEvents.CalendarEventDeleted]: (data) => this.calendarEventHandler.calendarEventDeleted(data),
            [api_1.WebSocketEvents.CalendarEventUpdated]: (data) => this.calendarEventHandler.calendarEventUpdated(data),
            [api_1.WebSocketEvents.CalendarEventRsvpUpdated]: (data) => this.calendarEventRsvpHandler.calendarEventRsvpUpdated(data),
            [api_1.WebSocketEvents.CalendarEventRsvpManyUpdated]: (data) => this.calendarEventRsvpHandler.calendarEventRsvpManyUpdated(data),
            [api_1.WebSocketEvents.CalendarEventRsvpDeleted]: (data) => this.calendarEventRsvpHandler.calendarEventRsvpDeleted(data),
            [api_1.WebSocketEvents.ChatMessageCreated]: (data) => this.messageHandler.messageCreated(data),
            [api_1.WebSocketEvents.ChatMessageDeleted]: (data) => this.messageHandler.messageDeleted(data),
            [api_1.WebSocketEvents.ChatMessageUpdated]: (data) => this.messageHandler.messageUpdated(data),
            [api_1.WebSocketEvents.ServerMemberJoined]: (data) => this.ServerMemberHandler.serverMemberJoined(data),
            [api_1.WebSocketEvents.ServerMemberRemoved]: (data) => this.ServerMemberHandler.serverMemberRemoved(data),
            [api_1.WebSocketEvents.ServerMemberUpdated]: (data) => this.ServerMemberHandler.serverMemberUpdated(data),
            [api_1.WebSocketEvents.ServerRolesUpdated]: (data) => this.ServerHandler.serverRolesUpdated(data),
            [api_1.WebSocketEvents.ServerMemberBanned]: (data) => this.ServerMemberHandler.serverMemberBanned(data),
            [api_1.WebSocketEvents.ServerMemberUnbanned]: (data) => this.ServerMemberHandler.serverMemberUnbanned(data),
            [api_1.WebSocketEvents.ServerMemberSocialLinkCreated]: (data) => this.ServerMemberHandler.serverMemberSocialLinkCreated(data),
            [api_1.WebSocketEvents.ServerMemberSocialLinkUpdated]: (data) => this.ServerMemberHandler.serverMemberSocialLinkUpdated(data),
            [api_1.WebSocketEvents.ServerMemberSocialLinkDeleted]: (data) => this.ServerMemberHandler.serverMemberSocialLinkDeleted(data),
            [api_1.WebSocketEvents.ServerWebhookCreated]: (data) => this.ServerWebhookHandler.serverWebhookCreated(data),
            [api_1.WebSocketEvents.ServerWebhookUpdated]: (data) => this.ServerWebhookHandler.serverWebhookUpdated(data),
            [api_1.WebSocketEvents.BotServerMembershipCreated]: (data) => this.botHandler.botServerMembershipCreated(data),
            [api_1.WebSocketEvents.BotServerMembershipDeleted]: (data) => this.botHandler.botServerMembershipDeleted(data),
            [api_1.WebSocketEvents.ListItemCompleted]: (data) => this.listHandler.listItemCompleted(data),
            [api_1.WebSocketEvents.ListItemUncompleted]: (data) => this.listHandler.listItemUncompleted(data),
            [api_1.WebSocketEvents.ListItemCreated]: (data) => this.listHandler.listItemCreated(data),
            [api_1.WebSocketEvents.ListItemUpdated]: (data) => this.listHandler.listItemUpdated(data),
            [api_1.WebSocketEvents.ListItemDeleted]: (data) => this.listHandler.listItemDeleted(data),
            [api_1.WebSocketEvents.DocCreated]: (data) => this.docHandler.docCreated(data),
            [api_1.WebSocketEvents.DocDeleted]: (data) => this.docHandler.docDeleted(data),
            [api_1.WebSocketEvents.DocUpdated]: (data) => this.docHandler.docUpdated(data),
            [api_1.WebSocketEvents.ServerChannelCreated]: (data) => this.ServerChannelHandler.serverChannelCreated(data),
            [api_1.WebSocketEvents.ServerChannelDeleted]: (data) => this.ServerChannelHandler.serverChannelDeleted(data),
            [api_1.WebSocketEvents.ServerChannelUpdated]: (data) => this.ServerChannelHandler.serverChannelUpdated(data),
            [api_1.WebSocketEvents.ChannelMessageReactionCreated]: (data) => this.reactionHandler.messageReactionCreated(data),
            [api_1.WebSocketEvents.ChannelMessageReactionDeleted]: (data) => this.reactionHandler.messageReactionDeleted(data),
            [api_1.WebSocketEvents.ForumTopicCreated]: (data) => this.forumHandler.forumTopicCreated(data),
            [api_1.WebSocketEvents.ForumTopicDeleted]: (data) => this.forumHandler.forumTopicDeleted(data),
            [api_1.WebSocketEvents.ForumTopicUpdated]: (data) => this.forumHandler.forumTopicUpdated(data),
            [api_1.WebSocketEvents.ForumTopicPinned]: (data) => this.forumHandler.forumTopicPinned(data),
            [api_1.WebSocketEvents.ForumTopicUnpinned]: (data) => this.forumHandler.forumTopicUnpinned(data),
            [api_1.WebSocketEvents.ForumTopicLocked]: (data) => this.forumHandler.forumTopicLocked(data),
            [api_1.WebSocketEvents.ForumTopicUnlocked]: (data) => this.forumHandler.forumTopicUnlocked(data),
        };
    }
    handleWSMessage(event, data) {
        var _a;
        const discardEventOption = (_a = this.client.options.gateway) === null || _a === void 0 ? void 0 : _a.discardEvent;
        if (discardEventOption === null || discardEventOption === void 0 ? void 0 : discardEventOption(event, data))
            return;
        const existingEvent = this.eventToHandlerMap[event];
        if (existingEvent) {
            void existingEvent(data);
        }
        else {
            this.client.emit("unknownGatewayEvent", data);
        }
    }
}
exports.ClientGatewayHandler = ClientGatewayHandler;
