// guilded.js/dist/structures/Webhook.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Webhook = void 0;
const util_1 = require("../util");
const Base_1 = require("./Base");
/**
 * Object representing received webhook data. This object is NOT to be used to send data to webhooks. That would be WebhookClient
 */
class Webhook extends Base_1.Base {
    constructor(client, data) {
        super(client, data);
        /**
         * The date this webhook was deleted if it was deleted
         */
        this._deletedAt = null;
        this.serverId = data.serverId;
        this._createdAt = (0, util_1.parseToStamp)(data.createdAt);
        this.authorID = data.createdBy;
        this._update(data);
    }
    /**
     * Returns the creation date of this webhook
     *
     * @returns The creation date of this webhook
     */
    get createdAt() {
        return new Date(this._createdAt);
    }
    /**
     * Returns the deletion date of this webhook if it was deleted
     *
     * @returns The deletion date of this webhook if it was deleted, otherwise null
     */
    get deletedAt() {
        return this._deletedAt ? new Date(this._deletedAt) : null;
    }
    /**
     * Returns the author of this webhook
     *
     * @returns The author of this webhook, or null if the author is not cached
     */
    get user() {
        var _a;
        return (_a = this.client.users.cache.get(this.id)) !== null && _a !== void 0 ? _a : null;
    }
    /**
     * Updates this webhook with new options
     *
     * @param options The new options for this webhook
     * @returns A promise that resolves with the updated webhook
     */
    update(options) {
        return this.client.webhooks.update(this.serverId, this.id, options);
    }
    /**
     * Deletes this webhook
     *
     * @returns A promise that resolves with this webhook after it has been deleted
     */
    delete() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.webhooks.delete(this.serverId, this.id);
            return this;
        });
    }
    _update(data) {
        var _a;
        if ("name" in data && data.name !== undefined)
            this.name = data.name;
        if ("channelId" in data && data.channelId !== undefined)
            this.channelID = data.channelId;
        if ("token" in data && data.token !== undefined)
            this.token = (_a = data.token) !== null && _a !== void 0 ? _a : null;
        if ("deletedAt" in data && data.deletedAt !== undefined) {
            this._deletedAt = data.deletedAt ? (0, util_1.parseToStamp)(data.deletedAt) : null;
        }
        return this;
    }
}
exports.Webhook = Webhook;
